File: src/NetWasm.Hosting/Deployment/DeploymentManifestValidator.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace NetWasm.Hosting.Deployment;

/// <summary>Validates the publish-safe structure and canonical identities of a deployment manifest.</summary>
public sealed class DeploymentManifestValidator : IDeploymentManifestValidator
{
    public void Validate(DeploymentManifest manifest)
    {
        ArgumentNullException.ThrowIfNull(manifest);
        if (manifest.SchemaVersion != 1)
        {
            throw new ArgumentException("Unsupported NetWasm deployment manifest schema.", nameof(manifest));
        }

        ValidateDigest(manifest.SemanticBuildId);
        ValidateDigest(manifest.BuildFingerprint);
        if (!Enum.IsDefined(manifest.DeploymentKind))
        {
            throw new ArgumentException("Unsupported NetWasm deployment kind.", nameof(manifest));
        }

        if (!string.Equals(manifest.Profile, "netwasm0.1", StringComparison.Ordinal))
        {
            throw new ArgumentException("Unsupported NetWasm managed profile.", nameof(manifest));
        }

        if (manifest.Target is not ("wasm32" or "wasm64"))
        {
            throw new ArgumentException("Unsupported NetWasm target.", nameof(manifest));
        }

        ValidateText(manifest.FeatureSet);
        ValidateVersionedIdentifier(manifest.ExecutionContract);
        ValidateVersions(manifest.Versions);
        ValidateRuntimeFeatures(manifest.RuntimeFeatures);
        ValidateArtifacts(manifest.Artifacts, manifest.RuntimeFeatures.Contains("local-time", StringComparer.Ordinal));
        ValidateImportModules(manifest.RequiredImportModules, manifest.RequiredImports);
        ValidateFunctions(manifest.RequiredImports);
        ValidateFunctions(manifest.Exports);
    }

    private static void ValidateImportModules(
        ImmutableArray<string> modules,
        ImmutableArray<DeploymentFunction> functions)
    {
        if (modules.IsDefault)
        {
            throw new ArgumentException(
                "Deployment required import modules must be explicit.",
                nameof(modules));
        }

        var identities = new HashSet<string>(StringComparer.Ordinal);
        foreach (var module in modules)
        {
            ValidateVersionedIdentifier(module);
            if (!identities.Add(module))
            {
                throw new ArgumentException(
                    "Deployment required import modules must be unique.",
                    nameof(modules));
            }
        }

        if (!functions.IsDefault
            && functions.Any(function => function is not null
                && !identities.Contains(function.Interface)))
        {
            throw new ArgumentException(
                "Every required import function must belong to a required import module.",
                nameof(modules));
        }
    }

    private static void ValidateVersions(DeploymentVersions versions)
    {
        ArgumentNullException.ThrowIfNull(versions);
        ValidateText(versions.Sdk);
        ValidateText(versions.Compiler);
        ValidateText(versions.Runtime);
        ValidateText(versions.RuntimeAbi);
        ValidateText(versions.Hosting);
        ValidateText(versions.Toolchain);
    }

    private static void ValidateRuntimeFeatures(ImmutableArray<string> features)
    {
        if (features.IsDefault)
        {
            throw new ArgumentException("Deployment runtime features must be explicit.", nameof(features));
        }

        var identities = new HashSet<string>(StringComparer.Ordinal);
        foreach (var feature in features)
        {
            ValidateToken(feature);
            if (!string.Equals(feature, "local-time", StringComparison.Ordinal))
            {
                throw new ArgumentException("Unsupported deployment runtime feature.", nameof(features));
            }

            if (!identities.Add(feature))
            {
                throw new ArgumentException("Deployment runtime features must be unique.", nameof(features));
            }
        }
    }

    private static void ValidateArtifacts(
        ImmutableArray<DeploymentArtifact> artifacts,
        bool hasLocalTime)
    {
        if (artifacts.IsDefaultOrEmpty)
        {
            throw new ArgumentException("A deployment must contain its application artifact.", nameof(artifacts));
        }

        var paths = new HashSet<string>(StringComparer.Ordinal);
        var applicationCount = 0;
        DeploymentArtifact? application = null;
        DeploymentArtifact? timeZone = null;
        foreach (var artifact in artifacts)
        {
            ArgumentNullException.ThrowIfNull(artifact);
            ValidateRelativePath(artifact.RelativePath);
            if (!paths.Add(artifact.RelativePath))
            {
                throw new ArgumentException("Deployment artifact paths must be unique.", nameof(artifacts));
            }

            ValidateToken(artifact.Role);
            ValidateMediaType(artifact.MediaType);
            ValidateDigest(artifact.Sha256);
            if (artifact.SchemaVersion is <= 0)
            {
                throw new ArgumentException("Subordinate artifact schema versions must be positive.", nameof(artifacts));
            }

            if (string.Equals(artifact.Role, "application", StringComparison.Ordinal))
            {
                applicationCount++;
                application = artifact;
            }
            else if (string.Equals(artifact.Role, "timezone-data", StringComparison.Ordinal))
            {
                if (timeZone is not null)
                {
                    throw new ArgumentException("A deployment may contain only one timezone artifact.", nameof(artifacts));
                }

                timeZone = artifact;
            }
        }

        if (applicationCount != 1)
        {
            throw new ArgumentException("A deployment must identify exactly one application artifact.", nameof(artifacts));
        }

        if (timeZone is null)
        {
            return;
        }

        if (!hasLocalTime)
        {
            throw new ArgumentException("A timezone artifact requires the local-time runtime feature.", nameof(artifacts));
        }

        if (!string.Equals(timeZone.RelativePath, application!.RelativePath + ".tz-info", StringComparison.Ordinal)
            || !string.Equals(timeZone.MediaType, "application/octet-stream", StringComparison.Ordinal)
            || timeZone.SchemaVersion != 1)
        {
            throw new ArgumentException("The timezone artifact must use the adjacent schema-1 deployment contract.", nameof(artifacts));
        }
    }

    private static void ValidateFunctions(ImmutableArray<DeploymentFunction> functions)
    {
        if (functions.IsDefault)
        {
            throw new ArgumentException("Deployment function inventories must be explicit.", nameof(functions));
        }

        var identities = new HashSet<string>(StringComparer.Ordinal);
        foreach (var function in functions)
        {
            ArgumentNullException.ThrowIfNull(function);
            ValidateVersionedIdentifier(function.Interface);
            ValidateFunctionName(function.Name);
            if (!identities.Add(string.Concat(function.Interface, "\0", function.Name)))
            {
                throw new ArgumentException("Deployment interface functions must be unique.", nameof(functions));
            }

            ValidateSignature(function.Parameters);
            ValidateSignature(function.Results);
        }
    }

    private static void ValidateSignature(ImmutableArray<string> values)
    {
        if (values.IsDefault)
        {
            throw new ArgumentException("Deployment function signatures must be explicit.", nameof(values));
        }

        foreach (var value in values)
        {
            ValidateText(value);
        }
    }

    private static void ValidateDigest(string digest)
    {
        ArgumentNullException.ThrowIfNull(digest);
        if (digest.Length != 64 || !digest.All(char.IsAsciiHexDigitLower))
        {
            throw new ArgumentException("Deployment identities require lowercase SHA-256 digests.", nameof(digest));
        }
    }

    private static void ValidateRelativePath(string path)
    {
        ValidateText(path);
        if (path[0] == '/' || path.Contains('\\', StringComparison.Ordinal)
            || path.Contains(':', StringComparison.Ordinal))
        {
            throw new ArgumentException("Deployment artifact paths must use canonical relative slash syntax.", nameof(path));
        }

        var segments = path.Split('/');
        if (segments.Any(segment => segment.Length == 0 || segment is "." or ".."))
        {
            throw new ArgumentException("Deployment artifact paths must not contain empty or traversal segments.", nameof(path));
        }
    }

    private static void ValidateMediaType(string mediaType)
    {
        ValidateText(mediaType);
        if (!mediaType.Contains('/', StringComparison.Ordinal))
        {
            throw new ArgumentException("Deployment artifact media types must be explicit.", nameof(mediaType));
        }
    }

    private static void ValidateVersionedIdentifier(string value)
    {
        ValidateText(value);
        var separator = value.LastIndexOf('@');
        if (separator <= 0 || separator == value.Length - 1 || value.AsSpan(0, separator).Contains('@'))
        {
            throw new ArgumentException("Deployment contract and interface identities must include one exact version.", nameof(value));
        }

        foreach (var character in value.AsSpan(0, separator))
        {
            if (!(char.IsAsciiLetterLower(character) || char.IsAsciiDigit(character) || character is '-' or '.' or ':' or '/'))
            {
                throw new ArgumentException("Deployment contract and interface identities must use canonical lowercase syntax.", nameof(value));
            }
        }

        var versionParts = value[(separator + 1)..].Split('.');
        if (versionParts.Length != 3 || versionParts.Any(part => part.Length == 0 || !part.All(char.IsAsciiDigit)))
        {
            throw new ArgumentException("Deployment contract and interface identities require an exact numeric version.", nameof(value));
        }
    }

    private static void ValidateToken(string value)
    {
        ValidateText(value);
        if (!value.All(character => char.IsAsciiLetterLower(character) || char.IsAsciiDigit(character) || character is '-' or '.'))
        {
            throw new ArgumentException("Deployment tokens must use canonical lowercase syntax.", nameof(value));
        }
    }

    private static void ValidateFunctionName(string value)
    {
        if (!DeploymentFunctionName.IsCanonical(value))
        {
            throw new ArgumentException(
                "Deployment function names must use canonical WIT identity syntax.",
                nameof(value));
        }
    }

    private static void ValidateText(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value);
        if (value.AsSpan().Trim().Length != value.Length || value.Any(char.IsControl))
        {
            throw new ArgumentException("Deployment text must be canonical and contain no control characters.", nameof(value));
        }
    }
}

File: src/NetWasm.Hosting/JavaScript/deployment-manifest-validator.mjs
import { isCanonicalWitFunctionName } from "./canonical-wit-function-name.mjs";

const rootKeys = [
  "artifacts",
  "buildFingerprint",
  "deploymentKind",
  "executionContract",
  "exports",
  "featureSet",
  "profile",
  "requiredImportModules",
  "requiredImports",
  "runtimeFeatures",
  "schemaVersion",
  "semanticBuildId",
  "target",
  "versions",
];
const versionKeys = ["compiler", "hosting", "runtime", "runtimeAbi", "sdk", "toolchain"];
const artifactKeys = ["mediaType", "relativePath", "role", "schemaVersion", "sha256"];
const functionKeys = ["interface", "name", "parameters", "results"];
const deploymentKinds = new Set(["browser", "component", "raw"]);
const targets = new Set(["wasm32", "wasm64"]);
const supportedRuntimeFeatures = new Set(["local-time"]);
const digestPattern = /^[0-9a-f]{64}$/u;
const tokenPattern = /^[a-z0-9][a-z0-9.-]*$/u;
const versionedIdentifierPattern = /^[a-z0-9][a-z0-9.\-:/]*@[0-9]+\.[0-9]+\.[0-9]+$/u;

export function validateDeploymentManifest(manifest) {
  assertExactDataObject(manifest, rootKeys, "deployment manifest");
  if (manifest.schemaVersion !== 1) {
    throw new TypeError("deployment manifest schema is unsupported");
  }
  validateDigest(manifest.semanticBuildId, "semantic build identity");
  validateDigest(manifest.buildFingerprint, "build fingerprint");
  if (!deploymentKinds.has(manifest.deploymentKind)) {
    throw new TypeError("deployment kind is unsupported");
  }
  if (manifest.profile !== "netwasm0.1") {
    throw new TypeError("managed profile is unsupported");
  }
  if (!targets.has(manifest.target)) {
    throw new TypeError("deployment target is unsupported");
  }
  validateText(manifest.featureSet, "feature set");
  validateVersionedIdentifier(manifest.executionContract, "execution contract");

  const versions = validateVersions(manifest.versions);
  const runtimeFeatures = validateRuntimeFeatures(manifest.runtimeFeatures);
  const artifacts = validateArtifacts(
    manifest.artifacts,
    runtimeFeatures.includes("local-time"));
  const requiredImportModules = validateImportModules(manifest.requiredImportModules);
  const requiredImports = validateFunctions(manifest.requiredImports, "required imports");
  for (const required of requiredImports) {
    if (!requiredImportModules.includes(required.interface)) {
      throw new TypeError(
        `required import '${required.interface}.${required.name}' has no required module`);
    }
  }
  const exports = validateFunctions(manifest.exports, "exports");

  return Object.freeze({
    schemaVersion: manifest.schemaVersion,
    semanticBuildId: manifest.semanticBuildId,
    deploymentKind: manifest.deploymentKind,
    profile: manifest.profile,
    target: manifest.target,
    featureSet: manifest.featureSet,
    executionContract: manifest.executionContract,
    versions,
    buildFingerprint: manifest.buildFingerprint,
    runtimeFeatures,
    artifacts,
    requiredImportModules,
    requiredImports,
    exports,
  });
}

function validateImportModules(value) {
  if (!Array.isArray(value)) {
    throw new TypeError("deployment required import modules must be explicit");
  }
  const identities = new Set();
  const modules = value.map(module => {
    validateVersionedIdentifier(module, "required import module");
    if (identities.has(module)) {
      throw new TypeError(`required import module '${module}' is duplicated`);
    }
    identities.add(module);
    return module;
  });
  return Object.freeze(modules);
}

function validateVersions(value) {
  assertExactDataObject(value, versionKeys, "deployment versions");
  const result = {};
  for (const key of versionKeys) {
    validateText(value[key], `deployment version '${key}'`);
    result[key] = value[key];
  }
  return Object.freeze(result);
}

function validateRuntimeFeatures(value) {
  if (!Array.isArray(value)) {
    throw new TypeError("deployment runtime features must be explicit");
  }
  const identities = new Set();
  const result = value.map(feature => {
    validateToken(feature, "deployment runtime feature");
    if (!supportedRuntimeFeatures.has(feature)) {
      throw new TypeError(`deployment runtime feature '${feature}' is unsupported`);
    }
    if (identities.has(feature)) {
      throw new TypeError(`deployment runtime feature '${feature}' is duplicated`);
    }
    identities.add(feature);
    return feature;
  });
  return Object.freeze(result);
}

function validateArtifacts(value, hasLocalTime) {
  if (!Array.isArray(value) || value.length === 0) {
    throw new TypeError("deployment artifacts must contain an application");
  }
  const paths = new Set();
  let application;
  let applicationCount = 0;
  let timeZone;
  const artifacts = value.map(artifact => {
    assertExactDataObject(artifact, artifactKeys, "deployment artifact");
    validateRelativePath(artifact.relativePath);
    if (paths.has(artifact.relativePath)) {
      throw new TypeError(`deployment artifact path '${artifact.relativePath}' is duplicated`);
    }
    paths.add(artifact.relativePath);
    validateToken(artifact.role, "deployment artifact role");
    validateMediaType(artifact.mediaType);
    validateDigest(artifact.sha256, "deployment artifact digest");
    if (artifact.schemaVersion !== null
        && (!Number.isInteger(artifact.schemaVersion) || artifact.schemaVersion <= 0)) {
      throw new TypeError("deployment artifact schema version must be null or positive");
    }
    const snapshot = Object.freeze({
      relativePath: artifact.relativePath,
      role: artifact.role,
      mediaType: artifact.mediaType,
      sha256: artifact.sha256,
      schemaVersion: artifact.schemaVersion,
    });
    if (artifact.role === "application") {
      applicationCount++;
      application = snapshot;
    } else if (artifact.role === "timezone-data") {
      if (timeZone !== undefined) {
        throw new TypeError("deployment contains multiple timezone artifacts");
      }
      timeZone = snapshot;
    }
    return snapshot;
  });
  if (applicationCount !== 1) {
    throw new TypeError("deployment must identify exactly one application artifact");
  }
  if (timeZone !== undefined) {
    if (!hasLocalTime) {
      throw new TypeError("timezone artifact requires the local-time runtime feature");
    }
    if (timeZone.relativePath !== `${application.relativePath}.tz-info`
        || timeZone.mediaType !== "application/octet-stream"
        || timeZone.schemaVersion !== 1) {
      throw new TypeError("timezone artifact does not satisfy the adjacent schema-1 contract");
    }
  }
  return Object.freeze(artifacts);
}

function validateFunctions(value, label) {
  if (!Array.isArray(value)) {
    throw new TypeError(`deployment ${label} must be explicit`);
  }
  const identities = new Set();
  const functions = value.map(func => {
    assertExactDataObject(func, functionKeys, "deployment function");
    validateVersionedIdentifier(func.interface, "deployment function interface");
    validateFunctionName(func.name);
    const identity = `${func.interface}\u0000${func.name}`;
    if (identities.has(identity)) {
      throw new TypeError(`deployment function '${func.interface}.${func.name}' is duplicated`);
    }
    identities.add(identity);
    const parameters = validateSignature(func.parameters);
    const results = validateSignature(func.results);
    return Object.freeze({
      interface: func.interface,
      name: func.name,
      parameters,
      results,
    });
  });
  return Object.freeze(functions);
}

function validateSignature(value) {
  if (!Array.isArray(value)) {
    throw new TypeError("deployment function signature must be explicit");
  }
  const result = value.map(item => {
    validateText(item, "deployment function signature value");
    return item;
  });
  return Object.freeze(result);
}

function validateRelativePath(value) {
  validateText(value, "deployment artifact path");
  if (value.startsWith("/") || value.includes("\\") || value.includes(":")) {
    throw new TypeError("deployment artifact path must use canonical relative slash syntax");
  }
  if (value.split("/").some(segment => segment.length === 0 || segment === "." || segment === "..")) {
    throw new TypeError("deployment artifact path contains an empty or traversal segment");
  }
}

function validateMediaType(value) {
  validateText(value, "deployment artifact media type");
  if (!value.includes("/")) {
    throw new TypeError("deployment artifact media type must be explicit");
  }
}

function validateDigest(value, label) {
  if (typeof value !== "string" || !digestPattern.test(value)) {
    throw new TypeError(`${label} must be a lowercase SHA-256 digest`);
  }
}

function validateVersionedIdentifier(value, label) {
  if (typeof value !== "string" || !versionedIdentifierPattern.test(value)) {
    throw new TypeError(`${label} must be a canonical exact versioned identifier`);
  }
}

function validateToken(value, label) {
  if (typeof value !== "string" || !tokenPattern.test(value)) {
    throw new TypeError(`${label} must use canonical lowercase syntax`);
  }
}

function validateFunctionName(value) {
  if (!isCanonicalWitFunctionName(value)) {
    throw new TypeError("deployment function name must use canonical WIT identity syntax");
  }
}

function validateText(value, label) {
  if (typeof value !== "string" || value.length === 0 || value.trim() !== value
      || /[\u0000-\u001f\u007f]/u.test(value)) {
    throw new TypeError(`${label} must be canonical non-empty text`);
  }
}

function assertExactDataObject(value, keys, label) {
  if (value === null || typeof value !== "object" || Array.isArray(value)
      || Object.getOwnPropertySymbols(value).length !== 0) {
    throw new TypeError(`${label} is invalid`);
  }
  const prototype = Object.getPrototypeOf(value);
  const descriptors = Object.getOwnPropertyDescriptors(value);
  const actualKeys = Object.keys(descriptors).sort();
  if (prototype !== null && prototype !== Object.prototype
      || actualKeys.length !== keys.length
      || actualKeys.some((key, index) => key !== keys[index])
      || Object.values(descriptors).some(descriptor => !descriptor.enumerable || !("value" in descriptor))) {
    throw new TypeError(`${label} shape is invalid`);
  }
}
